File: pfm-ts-node-react/explore/Trie.hs
{-# LANGUAGE OverloadedStrings #-}

{-
   HTTP Route Trie in Haskell

   This shows how naturally Tries work in Haskell with:
   - Immutable data structures by default
   - Pattern matching
   - Maybe monad for optional values
   - foldl for traversal

stack ghci --package pretty-simple --package containers ./Trie.hs
ghci> import Text.Pretty.Simple (pPrint)
ghci> emptyNode & addMethod "GET" & pPrint
ghci> pPrint buildSampleTrie

---

Or:

stack ghci --ghci-options "-interactive-print=Text.Pretty.Simple.pPrint" --package pretty-simple --package containers ./Trie.hs

-}

import Data.Function ((&))
import Data.Map (Map)
import Data.Map qualified as Map
import Data.Maybe (fromMaybe, isJust)
import Data.Set (Set)
import Data.Set qualified as Set

-- The Trie node type - algebraic data types make this clean
data TrieNode = TrieNode
  { children :: Map String TrieNode -- Map from segment to child
  , methods :: Set String -- HTTP methods at this node
  }
  deriving (Show, Eq)

-- State during traversal
data TrieState = TrieState
  { currentNode :: Maybe TrieNode -- Nothing means "not found"
  , capturedParams :: [String] -- Parameters from ? wildcards
  }
  deriving (Show, Eq)

-- Create an empty node
emptyNode :: TrieNode
emptyNode = TrieNode Map.empty Set.empty

-- Add a child to a node (returns new node - immutable!)
addChild :: String -> TrieNode -> TrieNode -> TrieNode
addChild segment child node =
  node{children = Map.insert segment child (children node)}

-- Add a method to a node
addMethod :: String -> TrieNode -> TrieNode
addMethod method node =
  node{methods = Set.insert method (methods node)}

-- The fold function - applied at each navigation step
-- This is the heart of the algorithm!
navigationStep :: TrieState -> String -> TrieState
navigationStep state segment =
  case currentNode state of
    Nothing ->
      -- Already at dead end, stay there
      state
    Just node ->
      -- First try exact match
      case Map.lookup segment (children node) of
        Just child ->
          -- Found exact match
          TrieState (Just child) (capturedParams state)
        Nothing ->
          -- No exact match, try wildcard
          case Map.lookup "?" (children node) of
            Just child ->
              -- Found wildcard, capture the parameter
              TrieState (Just child) (segment : capturedParams state)
            Nothing ->
              -- No match at all
              TrieState Nothing (capturedParams state)

-- Navigate through the trie using foldl'
navigate :: TrieNode -> [String] -> TrieState
navigate root segments =
  let initialState = TrieState (Just root) []
   in foldl navigationStep initialState segments

-- Build a sample trie
buildSampleTrie :: TrieNode
buildSampleTrie =
  let
    -- Create /users node with GET method
    usersNode =
      emptyNode
        & addMethod "GET"

    -- Create /users/? wildcard node with GET and DELETE

    -- Update users node to have wildcard child
    usersWithWildcard =
      (emptyNode & addMethod "GET") & (emptyNode & addMethod "GET" & addMethod "DELETE" & addChild "?")

    -- Build root with all paths
    root =
      emptyNode
        & addChild
          "users"
          ( -- /users node with GET method
            emptyNode
              & addMethod "GET"
              & ( -- /users/? wildcard node with GET and DELETE
                  emptyNode & addMethod "GET" & addMethod "DELETE" & addChild "?"
                )
          )
        & addChild "posts" (emptyNode & addMethod "GET") -- Create /posts node with GET method
   in
    root

-- Build the same trie using normal function application (no pipeline operators)
buildSampleTrie2 :: TrieNode
buildSampleTrie2 =
  let
    -- Create /users/? wildcard node with GET and DELETE methods
    userIdNode = addMethod "DELETE" (addMethod "GET" emptyNode)

    -- Create /users node with GET method and wildcard child
    usersNode = addChild "?" userIdNode (addMethod "GET" emptyNode)

    -- Create /posts node with GET method
    postsNode = addMethod "GET" emptyNode

    -- Build root with both children
    root = addChild "posts" postsNode (addChild "users" usersNode emptyNode)
   in
    root

buildSampleTrie3 :: TrieNode
buildSampleTrie3 =
  addChild
    "posts"
    (addMethod "GET" emptyNode)
    ( addChild
        "users"
        ( addChild
            "?"
            (addMethod "DELETE" (addMethod "GET" emptyNode))
            (addMethod "GET" emptyNode)
        )
        emptyNode
    )

{- FOURMOLU_DISABLE -}
buildSampleTrie4 :: TrieNode
buildSampleTrie4 =
  emptyNode
    & addChild "users"
        ( emptyNode
            & addMethod "GET"
            & addChild "?"
                ( emptyNode
                    & addMethod "GET"
                    & addMethod "DELETE"
                )
        )
    & addChild "posts"
        ( emptyNode
            & addMethod "GET"
        )
    & addChild "hello"
        ( emptyNode
            & addChild "?"
                ( emptyNode
                    & addMethod "GET"
                )
        )
{- FOURMOLU_ENABLE -}

-- Demonstrate fold iterations step by step
demonstrateFoldIterations :: IO ()
demonstrateFoldIterations = do
  let root = buildSampleTrie

  putStrLn "=== Haskell Trie Navigation Demo ===\n"

  -- Show iteration 0: Start at root
  let state0 = TrieState (Just root) []
  putStrLn $ "Iteration 0 (initial): node=" ++ show state0 ++ ", params=[]"

  -- Iteration 1: Navigate to "users"
  let state1 = navigationStep state0 "users"
  putStrLn $ "Iteration 1 (users):   node=" ++ show state1 ++ ", params=" ++ show (capturedParams state1)

  -- Iteration 2: Navigate to "123" (matches wildcard)
  let state2 = navigationStep state1 "123"
  putStrLn $ "Iteration 2 (123):     node=" ++ show state2 ++ ", params=" ++ show (reverse $ capturedParams state2)

  -- Check if GET method exists at final node
  case currentNode state2 of
    Just node
      | Set.member "GET" (methods node) ->
          putStrLn "✓ GET /users/123 found!"
    Just _ ->
      putStrLn "✗ No GET method at this path"
    Nothing ->
      putStrLn "✗ Path not found"

  putStrLn "\n=== Using foldl' (the natural way) ===\n"

  -- Test various paths
  let testPaths =
        [ (["users", "123"], "GET")
        , (["users"], "GET")
        , (["posts"], "GET")
        , (["posts", "456"], "GET") -- This won't match
        ]

  mapM_ (testPath root) testPaths
 where
  showNode :: TrieState -> String
  showNode state = case currentNode state of
    Just _ -> "Some(...)"
    Nothing -> "Nothing"

  testPath :: TrieNode -> ([String], String) -> IO ()
  testPath root (segments, method) = do
    let finalState = navigate root segments
        path = "/" ++ concatMap (++ "/") (init segments) ++ last segments
    case currentNode finalState of
      Just node
        | Set.member method (methods node) ->
            putStrLn $
              "✓ "
                ++ method
                ++ " "
                ++ path
                ++ " -> Found! Params: "
                ++ show (reverse $ capturedParams finalState)
      Just _ ->
        putStrLn $ "✗ " ++ method ++ " " ++ path ++ " -> Path found but no " ++ method ++ " method"
      Nothing ->
        putStrLn $ "✗ " ++ method ++ " " ++ path ++ " -> Path not found"

-- Alternative: Show the power of pattern matching with guards
matchRoute :: TrieNode -> String -> [String] -> Maybe [String]
matchRoute root method segments =
  case navigate root segments of
    TrieState (Just node) params
      | Set.member method (methods node) -> Just (reverse params)
      | otherwise -> Nothing
    _ -> Nothing

-- Main function
main :: IO ()
main = do
  demonstrateFoldIterations

  putStrLn "\n=== Pattern Matching Style ===\n"

  let root = buildSampleTrie

  -- Using pattern matching to handle results
  case matchRoute root "GET" ["users", "456"] of
    Just params -> putStrLn $ "Matched! Captured params: " ++ show params
    Nothing -> putStrLn "No match"

  case matchRoute root "DELETE" ["users", "789"] of
    Just params -> putStrLn $ "Matched! Captured params: " ++ show params
    Nothing -> putStrLn "No match"

File: pfm-ts-node-react/explore/bin/trie.ml
(* 
   HTTP Route Trie in OCaml
   
   This shows how a Trie naturally works in a functional language.
   Everything is immutable by default, pattern matching is built-in,
   and the fold pattern is idiomatic.
*)

module StringMap = Map.Make(String)
module StringSet = Set.Make(String)

(* The Trie node type - much cleaner than TypeScript! *)
type trie_node = {
  children : trie_node StringMap.t;  (* Map from segment to child node *)
  methods : StringSet.t;              (* Set of HTTP methods at this node *)
}

(* State during traversal - like our TrieStateFP *)
type trie_state = {
  current_node : trie_node option;    (* None means "not found" *)
  captured_params : string list;      (* Parameters captured from ? wildcards *)
}

(* Create an empty node *)
let empty_node = {
  children = StringMap.empty;
  methods = StringSet.empty;
}

(* Add a child to a node - returns new node (immutable!) *)
let add_child node segment child_node =
  { node with children = StringMap.add segment child_node node.children }

(* Add a method to a node *)
let add_method node method_name =
  { node with methods = StringSet.add method_name node.methods }

(* 
   The fold function - this is what gets applied at each step!
   In OCaml, this is much cleaner than TypeScript.
*)
let navigation_step (state : trie_state) (segment : string) : trie_state =
  match state.current_node with
  | None -> 
      (* Already at dead end, stay there *)
      state
  | Some node ->
      (* First try exact match *)
      match StringMap.find_opt segment node.children with
      | Some child -> 
          { current_node = Some child; 
            captured_params = state.captured_params }
      | None ->
          (* No exact match, try wildcard '?' *)
          match StringMap.find_opt "?" node.children with
          | Some child ->
              { current_node = Some child;
                captured_params = segment :: state.captured_params }
          | None ->
              (* No match at all *)
              { current_node = None; 
                captured_params = state.captured_params }

(* Navigate through the trie using fold - this is the key insight! *)
let navigate root segments =
  let initial_state = { current_node = Some root; captured_params = [] } in
  List.fold_left navigation_step initial_state segments

(* Build a sample trie *)
let build_sample_trie () =
  let root = empty_node in
  
  (* Add /users *)
  let users_node = add_method empty_node "GET" in
  let root = add_child root "users" users_node in
  
  (* Add /users/? (wildcard for user ID) *)
  let user_id_node = 
    empty_node 
    |> (fun n -> add_method n "GET")
    |> (fun n -> add_method n "DELETE") in
  let users_node = add_child users_node "?" user_id_node in
  let root = { root with children = StringMap.add "users" users_node root.children } in
  
  (* Add /posts *)
  let posts_node = add_method empty_node "GET" in
  let root = add_child root "posts" posts_node in
  
  root

(* Test the fold iterations step by step *)
let demonstrate_fold_iterations () =
  let root = build_sample_trie () in
  
  print_endline "=== OCaml Trie Navigation Demo ===\n";
  
  (* Show iteration 0: Start at root *)
  let state0 = { current_node = Some root; captured_params = [] } in
  Printf.printf "Iteration 0 (initial): node=%s, params=%s\n" 
    (if Option.is_some state0.current_node then "Some(root)" else "None")
    (String.concat ", " state0.captured_params);
  
  (* Iteration 1: Navigate to "users" *)
  let state1 = navigation_step state0 "users" in
  Printf.printf "Iteration 1 (users):   node=%s, params=%s\n"
    (if Option.is_some state1.current_node then "Some(users)" else "None")
    (String.concat ", " state1.captured_params);
  
  (* Iteration 2: Navigate to "123" (matches wildcard) *)
  let state2 = navigation_step state1 "123" in
  Printf.printf "Iteration 2 (123):     node=%s, params=[%s]\n"
    (if Option.is_some state2.current_node then "Some(?)" else "None")
    (String.concat ", " (List.rev state2.captured_params));
  
  (* Check if GET method exists at final node *)
  (match state2.current_node with
  | Some node -> 
      if StringSet.mem "GET" node.methods then
        print_endline "✓ GET /users/123 found!"
      else
        print_endline "✗ No GET method at this path"
  | None -> 
      print_endline "✗ Path not found");
  
  print_endline "\n=== Using fold_left (the natural way) ===\n";
  
  (* The beautiful part: using fold_left naturally! *)
  let test_paths = [
    (["users"; "123"], "GET");
    (["users"], "GET");
    (["posts"], "GET");
    (["posts"; "456"], "GET");  (* This won't match *)
  ] in
  
  List.iter (fun (segments, method_name) ->
    let final_state = navigate root segments in
    let path = "/" ^ String.concat "/" segments in
    match final_state.current_node with
    | Some node when StringSet.mem method_name node.methods ->
        Printf.printf "✓ %s %s -> Found! Params: [%s]\n" 
          method_name path 
          (String.concat ", " (List.rev final_state.captured_params))
    | Some _ ->
        Printf.printf "✗ %s %s -> Path found but no %s method\n" 
          method_name path method_name
    | None ->
        Printf.printf "✗ %s %s -> Path not found\n" method_name path
  ) test_paths

(* Run the demonstration *)
let () = demonstrate_fold_iterations ()